File: 0004-median-of-two-sorted-arrays/solution.java
class Solution {
    public double findMedianSortedArrays(int[] nums1, int[] nums2) {
        int n = nums1.length;
        int m = nums2.length;
        int s = n+m;

        int m1 = -1, m2 = -1;
        if((s & 1) == 1){
            m1= s/2;
            m2= s/2;
        }else{
            m1= s/2-1;
            m2= s/2;
        }

        int i = 0, j=0, cnt = 0;
        int median1 = Integer.MIN_VALUE, median2 = Integer.MIN_VALUE;

        while(i<n && j<m){
            if(median1 != Integer.MIN_VALUE  && median2 != Integer.MIN_VALUE){
                return (double)(median1 + median2)/2;
            }

            if(nums1[i] < nums2[j]){
                if(cnt == m1) median1 = nums1[i];
                if(cnt == m2) median2 = nums1[i];

                cnt++;
                i++;
            }else{
                if(cnt == m1) median1 = nums2[j];
                if(cnt == m2) median2 = nums2[j];

                cnt++;
                j++;
            }
        }

        while(i<n){
            if(median1 != Integer.MIN_VALUE  && median2 != Integer.MIN_VALUE){
                return (double)(median1 + median2)/2;
            }

            if(cnt == m1) median1 = nums1[i];
            if(cnt == m2) median2 = nums1[i];

            cnt++;
            i++;
        }

        while(j<m){
            if(median1 != Integer.MIN_VALUE  && median2 != Integer.MIN_VALUE){
                return (double)(median1 + median2)/2;
            }

            if(cnt == m1) median1 = nums2[j];
            if(cnt == m2) median2 = nums2[j];

            cnt++;
            j++;
        }

        if(median1 != Integer.MIN_VALUE  && median2 != Integer.MIN_VALUE){
            return (double)(median1 + median2)/2;
        }

        return -1;
    }
}

File: 0004-median-of-two-sorted-arrays/solution.cpp
class Solution {
public:
    double findMedianSortedArrays(vector<int>& nums1, vector<int>& nums2) {
        int s1 = nums1.size();
        int s2 = nums2.size();

        int totalLen = s1 + s2;
        int m1Ind = totalLen / 2;
        int m2Ind = totalLen / 2;

        if (totalLen % 2 == 0)
            m2Ind = m1Ind - 1;

        double m1 = INT_MIN;
        double m2 = INT_MIN;

        int i = 0, j = 0, cnt = -1;

        while (i < s1 && j < s2) {
            cnt++;
            if (nums1[i] < nums2[j]) {
                if (cnt == m1Ind)
                    m1 = nums1[i];
                if (cnt == m2Ind)
                    m2 = nums1[i];
                i++;
            } else {
                if (cnt == m1Ind)
                    m1 = nums2[j];
                if (cnt == m2Ind)
                    m2 = nums2[j];
                j++;
            }

            if (m1 != INT_MIN && m2 != INT_MIN)
                return (m1 + m2) / 2;
        }

        while (i < s1) {
            cnt++;
            if (cnt == m1Ind)
                m1 = nums1[i];
            if (cnt == m2Ind)
                m2 = nums1[i];
            i++;

            if (m1 != INT_MIN && m2 != INT_MIN)
                return (m1 + m2) / 2;
        }

        while (j < s2) {
            cnt++;
            if (cnt == m1Ind)
                m1 = nums2[j];
            if (cnt == m2Ind)
                m2 = nums2[j];
            j++;

            if (m1 != INT_MIN && m2 != INT_MIN)
                return (m1 + m2) / 2;
        }

        return -1; // just for the compiler
    }
};
